include/envoy/http/async_client.h

```diff
@@ -50,18 +83,64 @@ class AsyncClient {
     virtual void cancel() PURE;
   };
 
+  /**
+   * An in-flight HTTP stream.
+   */
+  class Stream {
+  public:
+    virtual ~Stream() {}
+
+    /***
+     * Send headers to the stream.
+     * @param headers supplies the headers to send.
+     * @param end_stream supplies whether this is a header only request.
+     */
+    virtual void sendHeaders(HeaderMap& headers, bool end_stream) PURE;
+
+    /***
+     * Send data to the stream.
+     * @param data supplies the data to send.
+     * @param end_stream supplies whether this is the last data.
+     */
+    virtual void sendData(Buffer::Instance& data, bool end_stream) PURE;
```

Comment:
A few things:
1) The codec APIs generally follow the http/2 framing format. Since in that API, if you want to close the stream after headers are complete, but without data, you must send an empty data frame along w/ the end stream bit, this is what we do also, for better or worse. Feel free to update the docs to account for this.
2) There are various asserts in different places to account for programmer error, but the main problem with this is due to object lifetime. Since streams may get destroyed right away, you are not guaranteed to actually be able to check vs. just be dealing with memory corruption / crash situation, so in general we just have a contract that says that once certain events happen you can't use the object anymore. I've tried to be very clear about this in codec.h and filter.h, but let me know if it's not clear enough.